opacus/optimizers/optimizer.py

```diff
@@ -8,19 +8,60 @@
 
 
 def _generate_noise(
-    std: float, reference: torch.Tensor, generator=None
+    std: float,
+    reference: torch.Tensor,
+    generator=None,
+    secure_mode: bool = False,
 ) -> torch.Tensor:
-    if std > 0:
-        # TODO: handle device transfers: generator and reference tensor
-        # could be on different devices
+    """
+    Generates noise according to a Gaussian distribution with mean 0
+
+    Args:
+        std: Standard deviation of the noise
+        reference: The reference Tensor to get the appripriate shape and device
+            for generating the noise
+        generator: The PyTorch noise generator
+        secure_mode: boolean showing if "secure" noise need to be generate
+            (see the notes)
+
+    Notes:
+        If `secure_mode` is enabled, the generated noise is also secure
+        against the floating point representation attacks, such as the ones
+        in https://arxiv.org/abs/2107.10138. This is achieved through calling
+        the Gaussian noise function 2*n times, when n=2 (see section 5.1 in
```

Comment:
curious, why is `n` set to 2?

Reply:
The trick that the distribution of the squared norm of a Gaussian is exponential. (see second sentence in "Gamma, exponential and related" in https://en.wikipedia.org/wiki/Chi-squared_distribution).
Thus, you can sample from a 2D gaussian by sampling first the norm (log(U) if U is uniform gives you exponential) and then the angle (uniform between 0 and 2Pi). That thus gives you two independent samples from the gaussian.

Reply:
So basically `n` can be any number > 1. The bigger, the more computation needs to be done (`2n` Gaussian samples will be generated). The reason we chose `n=2` is that, `n=1` could be easy to break and `n>2` is not really necessary. Let me explain: the complexity of the attack is `2^p(2n-1)`. In PyTorch, `p=53` and so complexity is `2^53(2n-1)`. With `n=1`, we get `2^53` (easy to break) but with `n=2`, we get `2^159`

Reply:
@alexandresablayrolles not sure if I understand what you explained, but I hope @karthikprasad got his response? :)

Reply:
> The trick that the distribution of the squared norm of a Gaussian is exponential. (see second sentence in "Gamma, exponential and related" in https://en.wikipedia.org/wiki/Chi-squared_distribution). Thus, you can sample from a 2D gaussian by sampling first the norm (log(U) if U is uniform gives you exponential) and then the angle (uniform between 0 and 2Pi). That thus gives you two independent samples from the gaussian.

@alexandresablayrolles You are right ... if the computations were done over the reals. Unfortunately, the two samples are *not* independent if computed with floating-point arithmetic. What we do instead is using the fact that a sum of Gaussians is a Gaussian, and transform two (or four) subtly correlated Gaussians into one with much smoother distribution.